lib/using_restricted_array.rb

```diff
@@ -6,58 +6,119 @@
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def length(array)
-  raise NotImplementedError
+  index = 0
+  while array[index] != nil do
+    index += 1
+  end
+  return index
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def print_array(array)
-  raise NotImplementedError
+  index = 0
+  while array[index] != nil do
+    puts array[index]
+    index += 1
+  end
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  raise NotImplementedError if length == 0
+  i = 0
+  while i < length do
+    if array[i] == value_to_find
+      return true
+    end
+    i += 1
+  end
+  return false
 end
 
+
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  raise NotImplementedError if length == 0
+  max_length = 0
+  length.times do |index|
+    if array[index] > max_length
+      max_length = array[index]
+    end
+  end
+  return max_length
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
+
+
 def find_smallest(array, length)
-  raise NotImplementedError
+  raise NotImplementedError if length == 0
+  min_length = array[0]
+  length.times do |index|
+    if array[index] < min_length
+      min_length = array[index]
+    end
+  end
+  return min_length
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n)
+# Space complexity: 0(1)
 def reverse(array, length)
-  raise NotImplementedError
+  raise NotImplementedError if length == 0
+  first_index = 0
+  last_index = length -1
+  while first_index < last_index
+    tamp = array[first_index]
+    array[first_index] = array[last_index]
+    array[last_index] = tamp
+
+    first_index += 1
+    last_index -= 1
+  end
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: 0(n) ?
```

Comment:
Actually because you cut the size of the to search array with each iteration the time complexity is O(log n).